### DISCUSSION TITLE
Fail captcha -> break

### DISCUSSION THREAD

**Discussion by Yurkhun** [thejoabo/virtualfisher-bot#69] [Ideas]
Hi,
I want to know if it's possible when the captcha verification fail instead of subtly getting a new captcha with /verify rege... get a break of some time and then try again because for what I see most of the time when I get the error of not getting answer from the ocr API it will fail all of the tries....

So instead of doing subtly the verify regen, the bot get a random break of maybe 120 - 600 seconds to make a new code.
Sometimes it seems OCR API is just not responding so it won't resolve any captcha and then it come back to work after a while 

And even if the bot miss all the 3/3 tries to verify and get the 15 minutes of wait to regen a new captcha it won't disconnect the bot... because it's happening very frequently to me to just disconnect the bot and I have to do again the command "python autofishbot.py" to restart the bot... so instead of disconnecting just put a "break" of 15 - 16 minutes in the bot and then it will automatically retry to do the captcha.

I wish I knew how to do those changes in the code 😅 but my knowledge of python is too little 🥲

**Comment by Yurkhun** [comment]
to explain it better... when I get that msg from virtualfisher bot "You have regenerated your captcha too many times in the past 15 minutes, you must wait before regenerating another."
the script instead of taking a break of 15 minutes it close automaticly and then I need to restart it manually... it would be amazing when it's not possible to regen a new captcha to just take a break then keep going later on
![WindowsTerminal_HQbVNFAQrD-min](https://user-images.githubusercontent.com/95500637/209359780-04e395ba-ffe4-4127-a8c0-957bf08e03ca.gif)

**Comment by thejoabo** [comment]
Hello @Yurkhun , This would be dangerous, if I'm not mistaken (and I read that a long time ago), captchas are limited to 7 attempts per captcha and limited to 5 regens. So if it fails to solve (due to external API errors) in the first three and it fails again users would be straight banned from VF. That's why I force manual input when **anything** goes wrong, to make it very clear to users that something is indeed, wrong. Automation is a tricky thing, specially in this case were there is multiple layers of bans, Discord itself and VF's internal service. 

There is probably a better solution, but I must put some thought to it before making any changes.

**Reply by Yurkhun** [reply]
Oh I see there is risk of ban 😅

**Reply by Yurkhun** [reply]
I thought it would be OK of after the 15 minutes the bot retry to do the captcha... but if there is ban it's kinda risk

**Reply by Maxhem2** [reply]
> Hello @Yurkhun , This would be dangerous, if I'm not mistaken (and I read that a long time ago), captchas are limited to 7 attempts per captcha and limited to 5 regens. So if it fails to solve (due to external API errors) in the first three and it fails again users would be straight banned from VF. That's why I force manual input when **anything** goes wrong, to make it very clear to users that something is indeed, wrong. Automation is a tricky thing, specially in this case were there is multiple layers of bans, Discord itself and VF's internal service.
> 
> There is probably a better solution, but I must put some thought to it before making any changes.

If this issue is still present it would be nice to get a push notification maybe to a webhook about the issue

**Reply by thejoabo** [reply]
> If this issue is still present it would be nice to get a push notification maybe to a webhook about the issue

Do you mean a push notification for what ? The external API online status ? Or for captcha-related problems in general ?

**Reply by Maxhem2** [reply]
This is my current solution, it does not restart the script to reattempt the captcha, but it messages the user via a webhook to get a little notification when having to verify manually. Replace YOUR_WEBHOOK_HERE with your webhook and place this script into the same folder as autofishbot.py is then start this script

```
import subprocess
import requests

# Execute the autofishbot.py script
subprocess.call(['python', 'autofishbot.py'])

# Send a message to the Discord webhook
webhook_url = "YOUR_WEBHOOK_HERE"
message = {
    'content': '@everyone The Auto Fishing Bot script has exited.',
    'avatar_url': 'https://i.ibb.co/Sc7fVcn/Download.png',
    'username': 'Auto Fishing Bot',
}
response = requests.post(webhook_url, json=message)
```

**Reply by thejoabo** [reply]
> This is my current solution, it does not restart the script to reattempt the captcha, but it messages the user via a webhook to get a little notification when having to verify manually. Replace YOUR_WEBHOOK_HERE with your webhook and place this script into the same folder as autofishbot.py is then start this script
> 
> ```
> import subprocess
> import requests
> 
> # Execute the autofishbot.py script
> subprocess.call(['python', 'autofishbot.py'])
> 
> # Send a message to the Discord webhook
> webhook_url = "YOUR_WEBHOOK_HERE"
> message = {
>     'content': '@everyone The Auto Fishing Bot script has exited.',
>     'avatar_url': 'https://i.ibb.co/Sc7fVcn/Download.png',
>     'username': 'Auto Fishing Bot',
> }
> response = requests.post(webhook_url, json=message)
> ```

Hello @Maxhem2 , this is a great solution for the "unexpected crash" alert. I was thinking of using regular OS notifications for unhandled/important events happening on the back-end, since I want to reduce at maximum the effort for users run the bot, and a OS notification (that doesn't require any user input - like creating this webhook) is better imo.

**Reply by Maxhem2** [reply]
> Hello @Maxhem2 , this is a great solution for the "unexpected crash" alert. I was thinking of using regular OS notifications for unhandled/important events happening on the back-end, since I want to reduce at maximum the effort for users run the bot, and a OS notification (that doesn't require any user input - like creating this webhook) is better imo.

While I understand the reasoning, a YouTube tutorial (which already are widely available explaining how to create a webhook) or similar would be more effective since ideally users would let the bot run on a vm in a datacenter which means os notification would never be read. Also your approach could prove troublesome for Linux users

**Reply by thejoabo** [reply]
> a YouTube tutorial (which already are widely available explaining how to create a webhook)

Agreed, I know how easy it is. But based on the number of people with errors getting their user-token and basic config stuff, I really want to make as plug-and-play as possible. 

> ideally users would let the bot run on a vm in a datacenter which means os notification would never be read

Agreed, I'm not so sure about the percentage of users that would really do that, but still. 

> Also your approach could prove troublesome for Linux users

Not really (I did some testing on debian-based and arch-based distros and it works fine), the most problematic one would be Mac. 

I mean, at the end of the day, I could just add another parameter in the config to enable either of them, but not as a requirement to run the bot. It's best to let users choose, and since it's not that hard to do both, that could be the case for these notifications.

**Comment by Yurkhun** [comment]
Hi guys, I found a solution to Auto Reconnect after the captcha fail and the bot disconnect..  whoever want to use here is the solution:
I made a change in the .bat/.cmd file I did previously
here the old .bat file for those who didn't see->[https://github.com/thejoabo/autofishbot/discussions/41#discussioncomment-4414805](https://github.com/thejoabo/autofishbot/discussions/41#discussioncomment-4414805)

The new code is:
```
@echo off 
title 1
python autofishbot.py 1
:a
cls
echo BOT RECONNECTING PLEASE WAIT!
timeout /t 1020 /nobreak
python autofishbot.py 1
goto a
```



to see how it works, here is the video showing it:

https://user-images.githubusercontent.com/95500637/209526808-a0761a72-0198-454f-9d43-27a7b9f6cf4d.mp4

**Reply by Yurkhun** [reply]
to change the timer, remember it's in seconds -> 900 seconds = 15 minutes

**Reply by Yurkhun** [reply]
but remember it might have a higher change to get banned...  I didn't get banned yet but always remember it can increase the chance
[https://github.com/thejoabo/autofishbot/discussions/69#discussioncomment-4488477](https://github.com/thejoabo/autofishbot/discussions/69#discussioncomment-4488477)

**Reply by Yurkhun** [reply]
Update -> instead of 900 seconds I put 1020 seconds... 2 minutes more for safety 😉 and a little visual changes with `@echo off` and` cls`

**Reply by Yurkhun** [reply]
This is the same reconnecting script for linux (.sh)

```
#!/bin/sh
python autofishbot.py 1
while :
do
  clear
  echo "RECONNECTING! PLEASE WAIT...."
  sleep 1020 | pv -t
  python autofishbot.py
done

**Comment by Maxhem2** [comment]
I am not sure if this feature is already implemented, but I don't believe it is. It would be beneficial if the captcha solver keeps sending requests to the API until it receives the same answer at least three times, and then use that answer. This would increase the likelihood of correctly solving captchas compared to using the first answer returned by the API.
Also Tesseract OCR might be a good option

**Reply by thejoabo** [reply]
> It would be beneficial if the captcha solver keeps sending requests to the API until it receives the same answer at least three times, and then use that answer.

It doesn't work that way. OCR.SPACE has 5 OCR engines, I send a request simultaneously to all of them. **Sending the same image to the same engine would lead to the same answer (either right or wrong) every time.**

Since the requests are sent **asynchronously**, they are also **filtered** (to only allow alphanumeric chars and also to test if the answer length is exactly 6) and if the result is valid, **tested** (this means, a "/verify {answer}" is sent to Discord) **asynchronously too** (all of it to solve it as fast as possible). It doesn't just test the first result at random, if only one engine produces a valid answer, only that one will be tested, on the other hand if 4 engines, for example, produces valid answers, it will test one by one (in the order that they were processed by the API) until it finds the correct one. 

And, as I said, there is no point sending that image to be analyzed again, therefore, if all tests fail, it sends a "/verify regen" and restarts the process.

> Also Tesseract OCR might be a good option

Already tested, it's not a good option, is way less reliable for captchas and would demand users to install heavy packages to run the bot (which is not ideal).
